### PR TITLE
[GEP-28] Add new e2e test skeleton for `gardenadm` functionality

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area dev-productivity
/kind enhancement

**What this PR does / why we need it**:

This PR adds make targets for the dev and e2e test setup for autonomous shoot clusters as well as the e2e test skeleton for `gardenadm`.

The setup is similar to the kind-based setup for gardener itself. We reuse the kind cluster, some skaffold configuration, and the provider-local machine pod image.

The e2e tests are executed on PRs and periodically on master (see https://github.com/gardener/ci-infra/pull/2827).
In contrast to the existing e2e tests, the new gardenadm tests use ginkgo's [ordered containers]() feature. I.e., there are individual `It` specs for each step instead of large `It` specs with multiple `By` steps.
This improves the test debugging and deflaking experience, as the output (also visualized in testgrid) points to an individual step where the tests failed.

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/2906

**Special notes for your reviewer**:

**Release note**:

```feature developer
A local setup for trying out, developing, and testing the autonomous shoot cluster functionality of `gardenadm` has been introduced. You can find the documentation [here](https://github.com/gardener/gardener/blob/master/docs/deployment/getting_started_locally_with_gardenadm.md).
```

### CODE REVIEW

**Comment by timebertt** [comment]
I added the missing documentation to this PR. It's ready for review now :)

**Comment by rfranzke** [comment]
@timebertt Thanks for the docs, I think it looks good (made minimal adjustments). I have fixed a gosec finding for the e2e tests - otherwise, I think the PR is ready to go. 
Looking forward to @timuthy's feedback!

**Review by marc1404 — COMMENTED** [review]
/lgtm

Thanks for the great documentation along with the test skeleton!
The getting started locally with `gardenadm` works out-of-the-box on my machine™ :)

File: docs/concepts/gardenadm.md [hunk 1]
```diff
@@ -8,6 +8,39 @@ description: Bootstrapping and management of autonomous shoot clusters.
 > Do not use it in production environments.
 > Read more about it in [GEP-28](../proposals/28-autonomous-shoot-clusters.md).
 
+<img src="../../logo/gardenadm-large.png" alt="gardenadm" width="100"/>
+
 ## Overview
 
-To be implemented.
+`gardenadm` is a command line tool for bootstrapping Kubernetes clusters called "Autonomous Shoot Clusters".
+In contrast to usual Gardener-managed clusters (called Shoot Clusters), the control plane runs as static pods on a dedicated control plane worker pool in the cluster itself (instead of running as pods on another Kubernetes cluster (called Seed Cluster)).
```
Comment (ScheererJ) [inline docs/concepts/gardenadm.md]: Nit: It depends what "control plane" means in this context. If you call all pods running in the shoot namespace on a seed "control plane" this statement may not be 100% correct as not all of those pods will run as static pods. That being said, just removing static is also not perfect.
In other words, as I have no better suggestion this can also stay as is. If you have an idea how to improve feel free to do so.

File: docs/concepts/gardenadm.md [hunk 2]
```diff
@@ -8,6 +8,39 @@ description: Bootstrapping and management of autonomous shoot clusters.
...
+Autonomous shoot clusters can be bootstrapped without an existing Gardener installation.
+Hence, they can host a Gardener installation itself and/or serve as the initial seed cluster of a Gardener installation.
+Furthermore, autonomous shoot clusters can only be created by the `gardenadm` tool and not via an API of an existing Gardener system.
+
+![Architectural overview](../proposals/assets/28-overview.png)
+
+Such autonomous shoot clusters are meant to operate autonomously, but not to exist completely independently of Gardener.
+Hence, after their initial creation, they are connected to an existing Gardener system such that the established cluster management functionality via the `Shoot` API can be applied.
+I.e., day-2 operations for autonomous shoot clusters are only supported after connecting them to a Gardener system.
```
Comment (ScheererJ) [inline docs/concepts/gardenadm.md:25]: Maybe we should add that the Gardener system may also run on the autonomous shoot cluster. It was mentioned before, but I would repeat it shortly here as well.

File: docs/concepts/gardenadm.md [hunk 3]
```diff
@@ -8,6 +8,39 @@ description: Bootstrapping and management of autonomous shoot clusters.
...
+
+Furthermore, autonomous shoot clusters are not considered a replacement or alternative for regular shoot clusters.
+They should be only used for special use-cases or requirements as their creation time will be longer and their costs will be higher.
```
Comment (ScheererJ) [inline docs/concepts/gardenadm.md]: Nit: Not sure if cluster creation time is really a KPI that is that important and that much different. The cost aspect seems to be based on assumptions that are not stated. Maybe we should add them. For example, costs may be higher as the nodes may not be utilised as much, but that is an assumption that does not necessarily hold true.

File: docs/concepts/gardenadm.md [hunk 4]
```diff
@@ -8,6 +8,39 @@ description: Bootstrapping and management of autonomous shoot clusters.
...
+The setup is also used for running e2e tests for `gardenadm` in CI.
+
+## Scenarios
+
+We distinguish between two different scenarios for bootstrapping autonomous shoot clusters:
+
+- [High Touch](#high-touch), meaning that there is no programmable infrastructure available.
+  We consider this the "bare metal" or "edge" use-case, where at first machines must be (often manually) prepared by human operators.
+  In this case, network setup (e.g., VPCs, subnets, route tables, etc.) and machine management are out of scope.
+- [Medium Touch](#medium-touch), meaning that there is programmable infrastructure available where we can leverage [provider extensions](../../extensions/README.md#infrastructure-provider) and [`machine-controller-manager`](https://github.com/gardener/machine-controller-manager) in order to manage the network setup and the machines.
```
Comment (ScheererJ) [inline docs/concepts/gardenadm.md]: Where should the high/medium touch links point to? There are no corresponding anchors in this file

File: docs/deployment/getting_started_locally_with_gardenadm.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+## High-Touch Scenario
+
+Use the following command to prepare the `gardenadm` high-touch scenario:
+
+```shell
+make gardenadm-high-touch-up
+```
+
+This will first build the needed images, deploy 2 machine pods using the [`gardener-extension-provider-local-node` image](../../pkg/provider-local/node), and install the `gardenadm` binary on both of them.
```
Comment (ScheererJ) [inline docs/deployment/getting_started_locally_with_gardenadm.md:55]: Just curious why you chose to use 2 machine pods? For highly available setups, 3 would be better. In a non-HA setup, the second machine can be used as a worker node, though.

File: docs/deployment/getting_started_locally_with_gardenadm.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+Afterward, you can use `kubectl exec` to execute `gardenadm` commands on the machines:
```
Comment (ScheererJ) [inline docs/deployment/getting_started_locally_with_gardenadm.md:57]: Nit: I am used to the British variant, but apparently you can also use "afterward" in American English.

```
Afterwards, you can use `kubectl exec` to execute `gardenadm` commands on the machines:
```

File: Makefile [hunk 7]
```diff
@@ -332,14 +332,14 @@ kind-operator-down: $(KIND)
...
+gardener%up gardener%dev gardener%debug gardenlet%up gardenlet%dev gardenlet%debug operator%up operator-dev operator-debug operator-seed-dev gardenadm%up: export SKAFFOLD_PUSH = true
+gardener%up gardener%dev gardener%debug gardenlet%up gardenlet%dev gardenlet%debug operator%up operator-dev operator-debug operator-seed-dev gardenadm%up: export SOURCE_DATE_EPOCH = $(shell date -d $(BUILD_DATE) +%s)
+gardener%up gardener%dev gardener%debug gardenlet%up gardenlet%dev gardenlet%debug operator%up operator-dev operator-debug operator-seed-dev gardenadm%up: export GARDENER_VERSION = $(VERSION)
 # use static label for skaffold to prevent rolling all gardener components on every `skaffold` invocation
-gardener%up gardener%dev gardener%debug gardener%down gardenlet%up gardenlet%dev gardenlet%debug gardenlet%down: export SKAFFOLD_LABEL = skaffold.dev/run-id=gardener-local
+gardener%up gardener%dev gardener%debug gardenlet%up gardenlet%dev gardenlet%debug operator%up gardenadm%up: export SKAFFOLD_LABEL = skaffold.dev/run-id=gardener-local
```
Comment (ScheererJ) [inline Makefile:340]: Why were `gardener%down` and `gardenlet%down` removed? Both do use skaffold as far as I can tell. Is this environment variable not needed there?

File: test/e2e/gardener/shoot/internal/rotation/ssh_keypair.go [hunk 8]
```diff
@@ -128,21 +127,18 @@ func (v *SSHKeypairVerifier) readAuthorizedKeysFile(ctx context.Context) (string
 		return "", fmt.Errorf("expected exactly one result when listing all machine pods: %+v", podList.Items)
 	}
 
-	podExecutor := kubernetes.NewPodExecutor(v.ShootFramework.SeedClient.RESTConfig())
-
-	reader, err := podExecutor.Execute(
+	stdout, _, err := v.ShootFramework.SeedClient.PodExecutor().Execute(
```
Comment (ScheererJ) [inline test/e2e/gardener/shoot/internal/rotation/ssh_keypair.go:130]: Not sure if we should really completely ignore `stderr` here. It is only a test, though.
WDYT?

File: test/framework/rootpod_executor.go [hunk 9]
```diff
@@ -72,17 +68,18 @@ func (e *rootPodExecutor) Execute(ctx context.Context, command string) ([]byte,
...
+	stdout, _, err := e.client.PodExecutor().Execute(ctx, e.Pod.Namespace, e.Pod.Name, e.Pod.Spec.Containers[0].Name,
+		append([]string{"chroot", "/hostroot"}, command...)...,
+	)
 	if err != nil {
-		if reader != nil {
-			response, readErr := io.ReadAll(reader)
+		if stdout != nil {
+			response, readErr := io.ReadAll(stdout)
 			return response, errors.Join(err, readErr)
 		}
```
Comment (ScheererJ) [inline test/framework/rootpod_executor.go]: This is not really doing the same thing as before. Previously, in case of error `stderr` was returned from the executor and evaluated here. Now, `stderr` is always ignored. Not sure if it was correct before, though.

Was this change intentional?

File: test/e2e/gardenadm/gardenadm.go [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/onsi/gomega/gbytes"
+)
+
+var _ = Describe("gardenadm Tests", Label("gardenadm", "default"), func() {
+	Describe("Single-node control plane", Ordered, Label("single"), func() {
+		It("should initialize the control plane node", func(ctx SpecContext) {
+			stdout, _, err := runtimeClient.PodExecutor().Execute(ctx, machinePod.Namespace, machinePod.Name, containerName, "gardenadm", "init")
```
Comment (ScheererJ) [inline test/e2e/gardenadm/gardenadm.go]: Do we not want to check `stderr` as well?

**Review by ScheererJ — COMMENTED** [review]
Thanks a lot for providing a great framework for testing GEP-28.

File: docs/deployment/getting_started_locally_with_gardenadm.md [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+The setup supports both the high-touch and medium-touch scenario of `gardenadm`.
+
+Based on [Skaffold](https://skaffold.dev/), the container images for all required components will be built and deployed into the cluster.
+This also includes the `gardenadm` CLI, which is installed on the machine pods by pulling the container image and extracting the binary.
+
+## Prerequisites
+
+- Make sure that you have followed the [Local Setup guide](../development/local_setup.md) up until the [Get the sources](../development/local_setup.md#get-the-sources) step.
+- Make sure your Docker daemon is up-to-date, up and running and has enough resources (at least `8` CPUs and `8Gi` memory; see [here](https://docs.docker.com/desktop/mac/#resources) how to configure the resources for Docker for Mac).
+  > Please note that 8 CPU / 8Gi memory might not be enough for more than two `Shoot` clusters, i.e., you might need to increase these values if you want to run additional `Shoot`s.
```
Comment (timuthy) [inline docs/deployment/getting_started_locally_with_gardenadm.md]: Is that information really meant for the `gardenadm` context? Seems like a copy-paste issue.

File: pkg/provider-local/node/certs.d/europe-docker.pkg.dev/hosts.toml [hunk 12]
```diff
@@ -0,0 +1,6 @@
+# europe-docker.pkg.dev upstream is required for loading the hyperkube image.
+# Further registries are supposed to be added via provider-local
+server = "http://europe.docker.pkg.dev"
```
Comment (timuthy) [inline pkg/provider-local/node/certs.d/europe-docker.pkg.dev/hosts.toml]: Do we really need this as part of the node image? It would be nice to still have the webook path being tested, i.e. not remove it from the `ensurer`.

File: test/e2e/gardenadm/gardenadm.go [hunk 13]
```diff
@@ -10,15 +10,43 @@ import (
...
-			stdout, _, err := runtimeClient.PodExecutor().Execute(ctx, machinePod.Namespace, machinePod.Name, containerName, "gardenadm", "init")
+			stdout, _, err := runtimeClient.PodExecutor().Execute(ctx, namespace, machinePodName(0), containerName,
+				"gardenadm", "init",
+			)
 			Expect(err).NotTo(HaveOccurred())
 
 			Eventually(ctx, gbytes.BufferReader(stdout)).Should(gbytes.Say("not implemented"))
 		}, SpecTimeout(5*time.Second))
+
+		It("should join the worker node", func(ctx SpecContext) {
```
Comment (timuthy) [inline test/e2e/gardenadm/gardenadm.go]:
```suggestion
		It("should join as worker node", func(ctx SpecContext) {
```

File: docs/deployment/getting_started_locally_with_gardenadm.md (hunk 5 above)
Comment (rfranzke) [reply docs/deployment/getting_started_locally_with_gardenadm.md:55]: HA is out of scope for now (we need to resolve the ETCD management questions first). The plan is to start with one control plane node and one worker node (which sums up to two).

File: Makefile (hunk 7 above)
Comment (rfranzke) [reply Makefile:340]: Yes, we indeed don't think it's needed for such rules and cleaned this up along the way.

File: test/e2e/gardener/shoot/internal/rotation/ssh_keypair.go (hunk 8 above)
Comment (rfranzke) [reply test/e2e/gardener/shoot/internal/rotation/ssh_keypair.go:130]: What do you propose to assert here?

File: test/framework/rootpod_executor.go (hunk 9 above)
Comment (rfranzke) [reply test/framework/rootpod_executor.go]: I don't remember whether it was intentional. Do have a concern and want to propose a change?

File: test/e2e/gardenadm/gardenadm.go (hunk 10 above)
Comment (rfranzke) [reply test/e2e/gardenadm/gardenadm.go]: At the moment, I don't think there is much benefit, or? What would you assert here?

File: pkg/provider-local/node/certs.d/europe-docker.pkg.dev/hosts.toml (hunk 12 above)
Comment (rfranzke) [reply pkg/provider-local/node/certs.d/europe-docker.pkg.dev/hosts.toml]: Nope, indeed we don't, see https://github.com/gardener/gardener/pull/11010. 

> It would be nice to still have the webook path being tested, i.e. not remove it from the ensurer.

We won't have anything to inject into this webhook anymore. This `europe.docker.pkg.dev` will be moved to the `reconcile` OSC only, and the `localhost` will become part of the `node` image.

File: test/e2e/gardener/shoot/internal/rotation/ssh_keypair.go (hunk 8 above)
Comment (ScheererJ) [reply test/e2e/gardener/shoot/internal/rotation/ssh_keypair.go:130]: I guess asserting that there is no `stderr` output would be what we usually expect. `gardenadm` should not randomly add some output on `stderr`.

File: test/e2e/gardenadm/gardenadm.go (hunk 10 above)
Comment (ScheererJ) [reply test/e2e/gardenadm/gardenadm.go]: Similar to https://github.com/gardener/gardener/pull/10977#discussion_r1883962889 I would suggest to check that `stderr` is empty.

File: test/framework/rootpod_executor.go (hunk 9 above)
Comment (ScheererJ) [reply test/framework/rootpod_executor.go]: The concern is that we might not see unexpected messages on `stderr`. Why not just add all `stderr` output as well?

Comment (rfranzke) [reply test/framework/rootpod_executor.go]: As discussed in person: I don't think this will be a problem, but let's await the test results.

**Review by ScheererJ — APPROVED** [review]
/lgtm
/approve

File: test/e2e/gardenadm/gardenadm.go (hunk 10 above)
Comment (rfranzke) [reply test/e2e/gardenadm/gardenadm.go]: Let's introduce this when it becomes relevant (i.e., when `gardenadm` really writes output to `stderr`).

File: docs/deployment/getting_started_locally_with_gardenadm.md [hunk 14]
```diff
@@ -25,9 +25,9 @@ This also includes the `gardenadm` CLI, which is installed on the machine pods b
 
 - Make sure that you have followed the [Local Setup guide](../development/local_setup.md) up until the [Get the sources](../development/local_setup.md#get-the-sources) step.
 - Make sure your Docker daemon is up-to-date, up and running and has enough resources (at least `8` CPUs and `8Gi` memory; see [here](https://docs.docker.com/desktop/mac/#resources) how to configure the resources for Docker for Mac).
-  > Please note that 8 CPU / 8Gi memory might not be enough for more than two `Shoot` clusters, i.e., you might need to increase these values if you want to run additional `Shoot`s.
-  Additionally, please configure at least `120Gi` of disk size for the Docker daemon.
-  > Tip: You can clean up unused data with `docker system df` and `docker system prune -a`.
+  > Additionally, please configure at least `120Gi` of disk size for the Docker daemon.
+  > [!TIP]
```
Comment (timebertt) [inline docs/deployment/getting_started_locally_with_gardenadm.md]: Nit: I don't think this markup works as the second line in a quote.

**Review by timebertt — COMMENTED** [review]
The new changes
/lgtm

**Review by timuthy — COMMENTED** [review]
Thank you!
/lgtm

**Review by timebertt — APPROVED** [review]
